pkg/handler/collector/oci/oci.go

```diff
@@ -0,0 +1,267 @@
+
+		tags, err := rc.TagList(ctx, r)
+		if err != nil {
+			return fmt.Errorf("reading tags for %s: %w", o.repo, err)
+		}
+
+		for _, tag := range tags.Tags {
+			if !strings.HasSuffix(tag, "sbom") && !strings.HasSuffix(tag, "att") && !strings.HasSuffix(tag, "sig") {
+				imageTag := fmt.Sprintf("%v:%v", o.repo, tag)
+				r, err := ref.New(imageTag)
+				if err != nil {
+					return err
+				}
+				err = o.fetchOCIArtifacts(ctx, rc, r, docChannel)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (o *ociCollector) fetchOCIArtifacts(ctx context.Context, rc *regclient.RegClient, image ref.Ref, docChannel chan<- *processor.Document) error {
+	logger := logging.FromContext(ctx)
+
+	// attempt to request only the headers, avoids Docker Hub rate limits
+	m, err := rc.ManifestHead(ctx, image)
+	if err != nil {
+		return err
+	}
+
+	if m.IsList() {
+		m, err := rc.ManifestGet(ctx, image)
+		if err != nil {
+			return err
+		}
+		pl, _ := manifest.GetPlatformList(m)
+		for _, p := range pl {
+			desc, err := manifest.GetPlatformDesc(m, p)
+			if err != nil {
+				return fmt.Errorf("failed retrieving platform specific digest: %w", err)
+			}
+			image.Digest = desc.Digest.String()
+			err = o.fetchOCIArtifacts(ctx, rc, image, docChannel)
+			if err != nil {
+				return fmt.Errorf("failed retrieving platform specific digest: %w", err)
+			}
+		}
+	}
+
+	digest := manifest.GetDigest(m)
+	digestFormatted := fmt.Sprintf("%v-%v", digest.Algorithm(), digest.Encoded())
+	// check to see if the digest has already been collected
+	if !contains(o.checkedDigest, digestFormatted) {
+		suffixList := []string{"att", "sbom"}
+		for _, suffix := range suffixList {
+			digestTag := fmt.Sprintf("%v.%v", digestFormatted, suffix)
+			imageTag := fmt.Sprintf("%v:%v", o.repo, digestTag)
+			r, err := ref.New(imageTag)
+			if err != nil {
+				return err
+			}
+
+			// if `.att` or `.sbom`` do not exist for specified digest
+			// log error and continue
+			m, err = rc.ManifestGet(ctx, r)
+			if err != nil {
+				logger.Error(err)
+				continue
+			}
+
+			// go through layers in reverse
+			mi, ok := m.(manifest.Imager)
+			if !ok {
+				return fmt.Errorf("reference is not a known image media type")
+			}
+			layers, err := mi.GetLayers()
+			if err != nil {
+				return err
+			}
+			for i := len(layers) - 1; i >= 0; i-- {
+				blob, err := rc.BlobGet(ctx, r, layers[i])
+				if err != nil {
+					return fmt.Errorf("failed pulling layer %d: %w", i, err)
+				}
+				btr1, err := blob.RawBody()
+				if err != nil {
+					return err
+				}
+
+				doc := &processor.Document{
+					Blob:   btr1,
+					Type:   processor.DocumentUnknown,
+					Format: processor.FormatUnknown,
+					SourceInformation: processor.SourceInformation{
+						Collector: string(OCICollector),
+						Source:    imageTag,
+					},
+				}
+				docChannel <- doc
+			}
+			o.checkedDigest = append(o.checkedDigest, fmt.Sprintf("%v-%v", digest.Algorithm(), digest.Encoded()))
```

Comment:
if an attestation/sbom doesn't exist the first time and then gets uploaded, it looks like it will miss it? 

also, should this include the repo as well?

Reply:
That is true for polling but the collector could be run again and collected.

Reply:
changing from  `checkDigest []string` to `checkedDigest map[string][]string` to include the repo along with the associated digests

Reply:
>That is true for polling but the collector could be run again and collected.

hmm would the `checkedDIgest` cache check would cause it not to look at the digest anymore after the first? Even if run again it would skip the digest?

Reply:
Only if a `sbom` or `att` is found it will be added to the `checkedDIgest` cache. So if it got added after, it would still get captured the next time around during polling. Sorry for the confusion.
```
			// if `.att` or `.sbom`` do not exist for specified digest
			// log error and continue
			m, err = rc.ManifestGet(ctx, r)
			if err != nil {
				logger.Error(err)
				continue
			}
```

Reply:
ah i mis-scoped that statement, i read it as after the main loop.

Reply:
should the cache check be on the included suffix level as well? (can we add a comment on assuming immutability of the metadata files?)

Reply:
Added digest+suffix to the cache